jobserver/api/releases.py

```diff
@@ -556,3 +558,73 @@ class WorkspaceStatusAPI(RetrieveAPIView):
 
     class serializer_class(serializers.Serializer):
         uses_new_release_flow = serializers.BooleanField()
+
+
+class TokenAuthenticationAPI(APIView):
+    authentication_classes = []
+
+    class serializer_class(serializers.Serializer):
+        user = serializers.CharField()
+        token = serializers.CharField()
+
+    class Level4AuthenticatedUser(serializers.Serializer):
+        username = serializers.CharField()
+        fullname = serializers.CharField()
+        workspaces = serializers.ListField(
+            child=serializers.CharField(),
+            default=[],
+        )
+        output_checker = serializers.BooleanField()
+        staff = serializers.BooleanField()
+
+    def post(self, request):
+        if getattr(request, "backend", None) is None:
+            logger.info("API Login with token failed because request not from backend")
+            raise NotAuthenticated()
+
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        data = serializer.data
+
+        try:
+            # user can be username or email
+            user = users.validate_login_token(data["user"], data["token"])
+        except (
+            User.DoesNotExist,
+            users.InvalidTokenUser,
+            users.BadLoginToken,
+            users.ExpiredLoginToken,
+        ) as exc:
+            logger.info(f"API Login with token failed for user {data['user']}: {exc}")
+            trace.get_current_span().record_exception(exc)
+            raise NotAuthenticated()
+
+        logger.info(f"User {user} logged in with login token via API")
+
+        workspaces = []
+        # this is 1 or 2 queries per project, not ideal, but we permissions are not stored in the db
```

Comment:
Yeah there's not much we can do about this given the current structure, is there? But as you say, given we only have to do it once per login it's hard to see it causing us too much trouble for a while.